### PR TITLE
Make ModuleResolutionKind.Node10 change backward-compatible

### LINKED ISSUE #53131: TS 5.0 - ModuleResolutionKind Enum changed (breaking change that's not listed)
# Bug Report

`ts.ModuleResolutionKind` seems to have changed between 4.9 and 5. The enum now has a `Node10` value which I guess is the old `NodeJs`. This isn't mentioned in the [API breaking changes](https://github.com/microsoft/TypeScript/wiki/API-Breaking-Changes) list, so this is either a bug (which will probably break a few libraries [for example the rollup typescript plugin, through which I discovered this]) or should be added to that list.

### 🔎 Search Terms

ModuleResolutionKind

### 🕗 Version & Regression Information

- This changed between versions 4.9 and 5.0

### ⏯ Playground Link

Doesn't apply here

### 💻 Code

```ts
ts.ModuleResolutionKind.NodeJs
```

### 🙁 Actual behavior

Error

### 🙂 Expected behavior

No error

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #53131

### CODE REVIEW

**Review by jakebailey — APPROVED** [review]

**Comment by jakebailey** [comment]
That CI result is... weird. is this somehow non-deterministic? Maybe we have to make `NodeJs` defined as `NodeJs = Node10`?

**Comment by andrewbranch** [comment]
No, I just forgot to run tests with type checking after I swapped the order.

**Comment by andrewbranch** [comment]
@typescript-bot cherry-pick to release-5.0

**Comment by jakebailey** [comment]
Oh, phew.

This reminds me that `Debug.formatEnum` always picks the _first_ name (so you can deprecate via an assignment), so this PR would actually debug print as the deprecated name (as that's first), but then the enum at runtime is the opposite. Unfortunate.

**Comment by andrewbranch** [comment]
The order has to be this way to make the enum reverse mapping point to the non-deprecated name, which appears in `tsc --traceResolution`. Honestly, `formatEnum` should probably do that too. Deprecated enum members that share a value with a non-deprecated member should always come first if reverse mapping ever matters.

**Comment by jakebailey** [comment]
Yeah, I agree. Unfortunately in SyntaxKind's case, we _must_ define old names in terms of new names (or vice versa) since the overall values are not stable (compared to this PR where we hardcoded "2").

But since it's vice versa, I can just send a PR that flips them around and it'll work.

**Comment by jakebailey** [comment]
Although, I think the reason to pick the first one is due to cases like:

```ts
enum NodeFlags {
    Number = 1 << 1,
    String = 1 << 2,
    Something = Number | String, // What if String is removed?
}
```

**Comment by andrewbranch** [comment]
I’m not sure why you can’t forward reference since it gets inlined anyway https://www.typescriptlang.org/play?#code/KYOwrgtgBAolDeAoKKoEEoF4oEYA0yqAQluogL5A

**Comment by jakebailey** [comment]
If you forward reference, you get an error:

> A member initializer in a enum declaration cannot reference members declared after it, including members defined in other enums. ts(2651)

Anyway, not super relevant to this PR besides `formatEnum` doing the wrong thing, but, I don't really think it's a big deal.

**Comment by andrewbranch** [comment]
Yeah, I meant I’m not sure why we issue that error in the first place.

**Comment by jakebailey** [comment]
Yeah, though if you add a forward reference, the value it emits is... not right.

- [playground](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBMAdgVwLZwKJwN4Cg4FwDCcAvHAIIA0+hFZcAjDYXAEIMW4C+QA)
- [esbuild](https://jakebailey.github.io/esbuild-playground/#Ly8gQGZpbGVuYW1lOiBjb25maWcuanNvbgp7CiAgImVudHJ5UG9pbnRzIjogWyIvaW5kZXgudHN4Il0sCiAgImZvcm1hdCI6ICJlc20iCn0KCi8vIEBmaWxlbmFtZTogaW5kZXgudHN4CmV4cG9ydCBlbnVtIEUgewogICAgQyA9IEEsCiAgICBBID0gMSwKICAgIEIgPSBBCn0=)

**Review by sheetalkamat — APPROVED** [review]

File: src/compiler/types.ts [hunk 1]
```diff
@@ -6940,6 +6940,12 @@ export function diagnosticCategoryName(d: { category: DiagnosticCategory }, lowe
 
 export enum ModuleResolutionKind {
     Classic  = 1,
+    /**
+     * @deprecated
+     * `NodeJs` was renamed to `Node10` to better reflect the version of Node that it targets.
+     * Use the new name or consider switching to a modern module resolution target.
+     */
+    NodeJs   = 2,
```
Comment (sheetalkamat) [inline src/compiler/types.ts:6948]: does this also means module: "nodejs" should also be done same way ?

Comment (sheetalkamat) [reply src/compiler/types.ts:6948]: Sorry got confused with "commonjs" moduleResolutionKind.

**Pull request merged**